lib/rex/post/postgresql/ui/console/command_dispatcher/client.rb

```diff
@@ -0,0 +1,140 @@
+
+            reqs = {}
+
+            filter_commands(cmds, reqs)
+          end
+
+          def name
+            'PostgreSQL Client'
+          end
+
+          def help_args?(args)
+            return false unless args.instance_of?(::Array)
+
+            args.include?('-h') || args.include?('--help')
+          end
+
+          def cmd_shell_help
+            print_line 'Usage: shell'
+            print_line
+            print_line 'Go into a raw SQL shell where SQL queries can be executed.'
+            print_line 'To exit, type `exit`, `quit`, `end` or `stop`.'
+            print_line
+          end
+
+          def cmd_shell(*args)
+            cmd_shell_help && return if help_args?(args)
+
+            prompt_proc_before = ::Reline.prompt_proc
+
+            ::Reline.prompt_proc = proc { |line_buffer| line_buffer.each_with_index.map { |_line, i| i > 0 ? 'SQL *> ' : 'SQL >> ' } }
+
+            stop_words = %w[stop s exit e end quit q].freeze
+
+            # Allow the user to query the DB in a loop.
+            finished = false
+            until finished
+              begin
+                # This will loop until it receives `true`.
+                raw_query = ::Reline.readmultiline('SQL >> ', use_history = true) do |multiline_input|
+                  finished = stop_words.include?(multiline_input.split.last)
+                  # Accept the input until the current line does not end with '\', similar to a shell
+                  finished || !multiline_input.split.last.end_with?('\\')
+                end
+              rescue ::Interrupt
+                finished = true
+              ensure
+                ::Reline.prompt_proc = prompt_proc_before
+              end
+
+              if finished
+                print_status 'Exiting Shell mode.'
+                return
+              end
+
+              formatted_query = raw_query.split.map { |word| word.chomp('\\') }.reject(&:empty?).compact.join(' ')
+
+              print_status "Running SQL Command: '#{formatted_query}'"
+              cmd_query(formatted_query)
+            end
+          end
+
+          def cmd_query_help
+            print_line 'Usage: query'
+            print_line
+            print_line 'You can also use `sql`.'
+            print_line 'Run a raw SQL query on the target.'
+            print_line
+          end
+
+          #
+          # @param [::Msf::Db::PostgresPR::Connection::Result] result The result of an SQL query to format.
+          def format_result(result)
+            columns = ['#']
+            columns.append(result.fields.map.each { |field| field[:name] })
+            flat_columns = columns.flatten
+
+            ::Rex::Text::Table.new(
+              'Header' => 'Query',
+              'Indent' => 4,
+              'Columns' => flat_columns,
+              'Rows' => result.rows.map.each.with_index do |row, i|
+                [i, row].flatten
+              end
+            )
+          end
+
+          def cmd_query(*args)
```

Comment:
When running an insert; msfconsole prints a half empty console - it'd be great to output a different message for that scenario

```
PostgreSQL @ 127.0.0.1:9000 (template1) > shell
SQL >> CREATE TABLE temp_table (data text);
[*] Running SQL Command: 'CREATE TABLE temp_table (data text);'
Query
=====

    #
    -

```

For instance, here's the equivalent result when using the `psql` utility:

```
$ psql postgres://postgres:password@127.0.0.1:9000         
psql (14.10 (Homebrew))
Type "help" for help.

postgres=# CREATE TABLE temp_table (data text);
CREATE TABLE
```

---

Inserts

```
SQL >> INSERT INTO temp_table VALUES ('dummy content');
[*] Running SQL Command: 'INSERT INTO temp_table VALUES ('dummy content');'
Query
=====

    #
    -
```

Versus psql

```
$ psql postgres://postgres:password@127.0.0.1:9000         
psql (14.10 (Homebrew))
Type "help" for help.

postgres=# CREATE TABLE temp_table (data text);
CREATE TABLE
postgres=# INSERT INTO temp_table VALUES ('dummy content');
INSERT 0 1
```